subprojects/docs/src/docs/userguide/upgrading_version_4.adoc

```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
+By definition this could impact the set of resolved dependencies of your build.
+However the fixed issues are mostly about corner cases and combination with recent features and thus should have a limited impact.
+
+When a dependency constraint matched a real dependency, it was made part of the graph.
+However if for some reason the dependency was later evicted from the graph, the constraint remained present. Now, when the last non-constraint edge to a dependency disappears, all constraints for that dependency will be properly removed from the graph.
+
+[[rel5.0:changes_to_default_task]]
+=== [5.0] Changes to property factory methods on DefaultTask
+
+Property factory methods on DefaultTask are final
+The property factory methods such as newInputFile() are intended to be called from the constructor of a type that extends DefaultTask. These methods are now final to avoid subclasses overriding these methods and using state that is not initialized.
+
+Inputs and outputs are not automatically registered
+The Property instances that are returned by these methods are no longer automatically registered as inputs or outputs of the task. The Property instances need to be declared as inputs or outputs in the usual ways, such as attaching annotations such as @OutputFile or using the runtime API to register the property.
+
+Previously:
+
+[source, groovy]
+----
+class MyTask extends DefaultTask {
+    // note: no annotation here
+    final RegularFileProperty outputFile = newOutputFile()
+}
+
+task myOtherTask {
+    def outputFile = newOutputFile()
+    doLast { ... }
+}
+----
+
+Now:
+
+[source, groovy]
+----
+class MyTask extends DefaultTask {
+    @OutputFile // property needs an annotation
+    final RegularFileProperty outputFile = project.objects.fileProperty()
+}
+
+task myOtherTask {
+    def outputFile = project.objects.fileProperty()
+    outputs.file(outputFile) // or to be registered using the runtime API
+    doLast { ... }
+}
+----
+
+[[rel5.0:jaxb_and_java9]]
+=== [5.0] Gradle now bundles JAXB for Java 9 and above
+
+In order to use S3 backed artifact repositories, it was previously required to add `--add-modules java.xml.bind` to `org.gradle.jvmargs` when running on Java 9 and above.
+
+Since Java 11 no longer contains the `java.xml.bind` module, Gradle now bundles JAXB 2.3.1 (`com.sun.xml.bind:jaxb-impl`) and uses it on Java 9 and above.
+
+Please remove the `--add-modules java.xml.bind` option from `org.gradle.jvmargs`, if set.
+
+[[rel5.0:java_library_distribution_plugin]]
+=== Java Library Distribution Plugin utilizes Java Library Plugin
+
+The <<#java_library_distribution_plugin, Java Library Distribution Plugin>> is now based on the
+<<#java_library_plugin, Java Library Plugin>> instead of the <<#java_plugin, Java Plugin>>.
+
+Additionally the created distribution will contain all artifacts of the `runtimeClasspath` configuration instead of the deprecated `runtime` configuration.
```

Comment:
"Additionally, the default distribution created by the plugin will contain...". The "created distribution" is a little unclear, although not particularly ambiguous once you think about it. But we shouldn't require our readers to think hard to understand what is meant.